lib/improver/wind_direction.py

```diff
@@ -0,0 +1,237 @@
+                angle = 0.0
+        return angle
+
+    @staticmethod
+    def find_r_values(complex_in):
+        """Takes input wind direction in complex values and returns array
+           containing r values using Pythagoras theorem."""
+
+        return np.sqrt(np.square(complex_in.real)+np.square(complex_in.imag))
+
+    @staticmethod
+    def calc_polar_mean_std_dev(wind_dir_complex, wind_dir_deg_mean):
+        """The averaged wind direction complex values have r values
+           between 0-1. So we need to recalculate the complex values so they
+           retain their angle but r is fixed as r=1. They can then be used to
+           find the distance between the average angle and the other
+           wind direction data points. From this information - we can calculate
+           a confidence value."""
+
+        wind_dir_complex_mean_fixr = WindDirection.deg_to_complex(
+                                                             wind_dir_deg_mean)
+
+        # Find difference in the distance between all the observed points and
+        # mean point with fixed r=1.
+        # For maths to work - the "wind_dir_complex_mean_fixr array" needs to
+        # be "tiled" so that it is the same dimension as "wind_dir_complex".
+        wind_dir_complex_mean_tile = np.tile(wind_dir_complex_mean_fixr,
+                                             (wind_dir_complex.shape[0], 1, 1))
+
+        # Calculate distance from each wind direction data point to the
+        # average point.
+        difference = wind_dir_complex - wind_dir_complex_mean_tile
+        dist_from_mean = np.sqrt(np.square(difference.real) +
+                                 np.square(difference.imag))
+
+        # Find average distance.
+        dist_from_mean_avg = np.mean(dist_from_mean, axis=0)
+
+        # If we have two points at opposite ends of the compass
+        # (eg. 270 and 90), then their separation distance is 2.
+        # Normalise the array using 2 as the maximum possible value.
+        dist_from_mean_norm = 1 - dist_from_mean_avg*0.5
+
+        return dist_from_mean_norm
+
+    @staticmethod
+    def wind_dir_decider(wind_dir_deg, wind_dir_deg_mean, r_vals):
+        """If the wind direction is so widely scattered that the r value
+           is nearly zero then this indicates that the average wind direction
+           is essentially meaningless.
+           We therefore subistuite this meaningless average wind
+           direction value for the wind direction taken from the first
+           ensemble member."""
+
+        # Threshold r value.
+        r_thresh = 0.01
+
+        # Mask True if r values below threshold.
+        where_low_r = np.where(r_vals < r_thresh, True, False)
+
+        # If the whole array contains good r-values, return origonal array.
+        if not where_low_r.any():
+            return wind_dir_deg_mean
+
+        # Takes first ensemble member.
+        first_member = wind_dir_deg[0]
```

Comment:
See my comment on assuming axis and coordinate order. It is slightly trickier here because you are working on an array not a cube, otherwise you would just take the first slice over the coordinate of interest (e.g. realization). You could pass cubes into these functions, and then just work on cube.data, but I'm not sure if that's the right approach. What do you think?

Reply:
It is tricky - I'm not sure of a good way of rebroadcasting the arrays without slowing down the code and making it more complex. The way that the code extracts the data should mean that the first dimension is always realisations.